[FEAT/#95] 그룹 참여 화면, Firestore 연결

TaewoongR opened the pull request:
- closed #95 

## *📍 Work Description*

- 그룹 참여 화면 제작
- Firestore 연결

## *📸 Screenshot*

https://github.com/user-attachments/assets/7146a8f3-8e8c-49e5-8421-f523d172a783

## *📢 To Reviewers*
- 추가한 주요 사항
  - core:model 모듈에 GroupModel 생성
  - data:mygroup 모듈 생성
  - domain:mybroup 모듈 생성
  - app 모듈 의존성에 data:mygroup 추가
  - settings.gradle.kts 에 생성한 두 모듈 include

- data 와 domain 모듈에서의 group 데이터 가져오는 방식은 동역님이 작성한 에피소드와 굉장히 유사합니다.
- FirebaseFirestore 인스턴스 주입은 기존 하나로 충분하여 data:episode 에 정의된 di module에서 객체를 주입받습니다.

- 아직 그룹 참여에는 firestore와 연결되어있지 않고 뼈대만 완성했습니다.

- - -

- 추가
  - 서버에서 모든 그룹을 가져오는 방식 -> 유저가 속해있는 그룹들만을 가져오는 방식

## ⏲️Time

    - ui 3시간 + firestore 연결 3시간

TaewoongR committed:
feat/#95 : 그룹 참여 ui 구성

- 초대코드 입력 전 화면

TaewoongR committed:
feat/#95 : 그룹 참여 ui 뼈대 완성

- 색상 속성 추가
- 커스텀 Divider 색상 변경

TaewoongR committed:
feat/#95 : 그룹 참여 ui 뼈대 디자인 수정

- 하단 버튼 디자인 패딩 설정

TaewoongR committed:
feat/#95 : firestore 연결

- data:mygroup 에서 firestore CRUD 작성
- firestore 용의 model, core:model 객체로 변환 함수 선언
- repositoryImpl 과 di module 정의
- domain:mygroup 에서 repository interface 정의
- viewmodel 에 주입하여 활용

shinythink commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/screen/GroupJoinScreen.kt:
```diff
@@ -1,17 +1,70 @@
 	onBackClick: () -> Unit,
 ) {
+	val focusManager = LocalFocusManager.current
+	var codeText by remember { mutableStateOf("") }
+	var isInput by remember { mutableStateOf(false) }
+	var isValid by remember { mutableStateOf(false) }
```
rememberSaveable 사용하시면 더 좋을 것 같습니다. 아직 viewModel이랑 연결이 되어 있지 않아서 임시로 둔 건가요??

TaewoongR replied:
이거 저번에 태희님이 언급해주셨는데 또 이렇게 했네요. 알려주셔서 감사해요!

shinythink commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/screen/GroupJoinScreen.kt:
```diff
@@ -29,5 +82,200 @@ fun GroupJoinScreen(
+				.padding(it)
+				.padding(horizontal = 20.dp),
+		) {
+			LazyColumn(
+				modifier = Modifier
+					.fillMaxSize(),
+				horizontalAlignment = Alignment.CenterHorizontally,
+			) {
```
LazyColumn으로 하신 이유가 있나요???

TaewoongR replied:
화면 회전 대응하려고 LazyColumn 사용했어요

shinythink commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/screen/GroupJoinScreen.kt:
```diff
@@ -29,5 +82,200 @@ fun GroupJoinScreen(
+		) {
+			AsyncImage(
+				model = "https://avatars.githubusercontent.com/u/127717111?v=4",
+				contentDescription = "",
```
나중엔 그룹이름을 넣게 되겠죠??

TaewoongR replied:
여기에는 그룹 이미지 url들어가면 됩니다.

shinythink commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/screen/GroupJoinScreen.kt:
```diff
@@ -29,5 +82,200 @@ fun GroupJoinScreen(
+					Spacer(modifier = Modifier.padding(top = 4.dp))
+					MapisodeDivider(direction = Direction.Horizontal, thickness = Thickness.Thin)
+					Spacer(modifier = Modifier.padding(5.dp))
+					MapisodeFilledButton(
+						onClick = {},
+						text = "참여하기",
+						showRipple = true,
+					)
```
초대 코드 입력 후 클릭하는 '확인' 버튼과 크기를 같게 하는 건 어떨까요???

TaewoongR replied:
네 반영해보겠음다

shinythink commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/screen/GroupJoinScreen.kt:
```diff
@@ -1,17 +1,70 @@
 ) {
+	val focusManager = LocalFocusManager.current
+	var codeText by remember { mutableStateOf("") }
+	var isInput by remember { mutableStateOf(false) }
```
지금 이름(IsInput)도 괜찮지만 좀 더 구체화해도 좋을 것 같습니다.

TaewoongR replied:
네 이름에서 좀 더 의도가 들어가면 좋아보이겠어요.

shinythinking reviewed (approved):
고생 많으셨습니다!!
주말에 같이 했는데 진도가 상당히 빠르네요 ㅠ
저도 분발하겠습니다!!!

ps) 유저 별 그룹 가져오는 것도 뚫으신다고 합니다!

Ameri-Kano commented on data/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/model/GroupModelConverter.kt:
```diff
@@ -0,0 +1,14 @@
+import com.google.firebase.Timestamp
+import com.google.firebase.firestore.FirebaseFirestore
+
+internal fun GroupModel.toFirestoreModel(database: FirebaseFirestore) = GroupFirestoreModel(
+	adminUser = database.collection("adminUser").document(adminUser),
+	createdAt = Timestamp(createdAt),
+	description = description,
+	imageUrl = imageUrl,
+	name = name,
+	members = members.map { database.collection("members").document(it) },
+)
```
adminUser, members 를 상수화하는 것도 좋아보입니다!
(지금은 core:firebase 의 FirestoreConstants 에 위치해 있습니다.)

TaewoongR replied:
나쁜 하드 코딩 습관 개선리뷰 감사해요

Ameri-Kano commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/intent/GroupState.kt:
```diff
@@ -1,36 +1,37 @@
-	val imageUrl: String,
-	val description: String,
-	val createdAt: String,
+data class GroupUiModel(
+	val id: String,
 	val adminUser: String,
-	val users: ImmutableList<User> = persistentListOf(),
+	val createdAt: String,
+	val description: String,
+	val imageUrl: String,
+	val name: String,
+	val members: List<String>,
 )
```
UiModel 에 `@Immutable` 을 붙이는 것은 Recomposition 과 관련이 있는 게 맞나요?

TaewoongR replied:
네 맞아요 다만 ImmutableList가 아니라 PersistentList로 해도 stable 인 점을 compose-compiler로 확인해서 수정할 예정입니다.

Ameri-Kano reviewed (approved):
주말에 수고 많으셨습니다!

haeti-dev commented on core/designsystem/src/main/kotlin/com/boostcamp/mapisode/designsystem/compose/MapisodeText.kt:
```diff
@@ -25,7 +25,7 @@ fun MapisodeText(
 	color: Color = Color.Unspecified,
 	onTextLayout: (TextLayoutResult) -> Unit = {},
 	style: MapisodeTextStyle = LocalTextStyle.current,
-	overflow: TextOverflow = TextOverflow.Clip,
```
기본값을 바꾼 경우에는 꼭 팀원들에게 공지 또는 논의 후 해주세요. 기존에 사용하던 다른 개발자 뷰에서 달라지면 안되니 소통을 더 적극적으로 하셔야 합니다!

TaewoongR replied:
굉장히 위험한 수정을 했었네요. 코딩할 때 좀 더 주의를 기울여야 겠어요

haeti-dev commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/intent/GroupState.kt:
```diff
@@ -1,36 +1,37 @@
 data class GroupState(
 	val areGroupsLoading: Boolean = false,
-	val groups: ImmutableList<GroupItemUiModel> = persistentListOf(),
+	val groups: ImmutableList<GroupUiModel> = persistentListOf(),
```
ImmutableList로 타입을 선언한 이유는 추가 삭제 연산을 사용하지 않을 예정이여서 그런건가요? PersistentList를 사용하지 않은 이유가 궁금합니다.

TaewoongR replied:
compose-compiler가 PersistentList를 불변으로 해석하지 못한다는 오해를 하고 Immutable로 변경했었습니다.
결과적으로는 PersistentList는 불변으로 해석해서 stable임을 확인했습니다.
unstable로 나오던 원인은 제너릭이 unstable 였기 때문이었습니다.

<img src="https://github.com/user-attachments/assets/fec19735-5c3f-4ba4-b712-9bc376c4da27" width=400>

그리고 wrapper class를 사용할 이유도 사라져서 곧바로 core:model 의 data class를 곧바로 사용할 것 같습니다.

haeti-dev reviewed (approved):
UI에 대해서 공통적으로 사용되는 컴포넌트는 해당 모듈 하위 폴더로 component 만들어서 정의하시는게 좋아보여요. 고생하셨습니다~

TaewoongR committed:
fix/#95 : repository 그룹 get 함수 수정

- 모든 그룹 -> 유저 id가 속한 그룹

TaewoongR committed:
fix/#95 : 코드 리뷰 반영

- immutableList -> persistentList 변경
- 그룹 참여 화면의 변수명 수정 : 의도 파악 쉽게 변경
- rememberSavable 사용으로 configuration change 대응

TaewoongR committed:
fix/#95 : for ktlint
